README.md
# vpnserver
Docker image for SoftEther VPN server. Small and ready to use.

This docker image is based on alpine base image. Total size is very small (about 14MB).

## Features

- Can be configured as a multi-hop vpn system.
- Can add more than one vpn user account.


## Contents

* `src` folder contains `genconfig` tool's source code.
* `docker` folder contains `Dockerfile` and it's dependents. 

## Config & Run

When first running, `genconfig` will config the vpnserver automatically.
For security reason, it will randomly generate server and virtual hub admin password.

The vpnserver is listening on port **8080** inside container.
You only need to set vpn user account and/or next hop vpn account (depends on whether you need multi-hop vpn system) in environment variable.

### **Single node (Standard) mode**

```shell
docker run -d -e UserList='user1:pass1;user2:pass2...' -p xxxx:8080 reaganer/vpnserver:latest
```

If `UserList` is not set or empty, `genconfig` will create an account with randomly generated username and password and print it on the `STDOUT`.

### **Relay node (Mutli-Hop vpn) mode**

The last hop node needs to be configured as **Single node mode**. Other nodes need to be configured as relay node mode.

You can use an env-file to configure all the environment variables at once.

```shell
$ cat node.env
RelayNode=TRUE
NextNodeHost=xxx.xxx.xxx.xxx
NextNodePort=xxxxx
NextNodeUserName=username
NextNodePassword=password
UserList=user1:pass1;user2:pass2...

$ docker run -d --env-file ./node.env -p xxxx:8080 reaganer/vpnserver:latest
```



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [2.1] - 2021-05-31

### Fixed

- README Mutli-Hop mode configuration error.

## [2.0] - 2020-06-03

### Added

- Mutli-Hop function.
- Mutli-User function.

### Changed

- Disable nat traversal function in config file.
- Change user config method.

## [1.0] - 2020-03-08

Initial version.

src/config.go
package main

const configTemplate = `
# Software Configuration File
# ---------------------------
# 
# You may edit this file when the VPN Server / Client / Bridge program is not running.
# 
# In prior to edit this file manually by your text editor,
# shutdown the VPN Server / Client / Bridge background service.
# Otherwise, all changes will be lost.
# 
declare root
{
	uint ConfigRevision 0
	bool IPsecMessageDisplayed true
	string Region US
	bool VgsMessageDisplayed true

	declare DDnsClient
	{
		bool Disabled true
		byte Key {{.DDnsClientKey}}
		string LocalHostname alpine
		string ProxyHostName $
		uint ProxyPort 0
		uint ProxyType 0
		string ProxyUsername $
	}
	declare IPsec
	{
		bool EtherIP_IPsec false
		string IPsec_Secret vpn
		string L2TP_DefaultHub DEFAULT
		bool L2TP_IPsec false
		bool L2TP_Raw false

		declare EtherIP_IDSettingsList
		{
		}
	}
	declare ListenerList
	{
		declare Listener0
		{
			bool DisableDos false
			bool Enabled true
			uint Port 8080
		}
	}
	declare LocalBridgeList
	{
		bool DoNotDisableOffloading false
	}
	declare ServerConfiguration
	{
		bool AcceptOnlyTls true
		uint64 AutoDeleteCheckDiskFreeSpaceMin 104857600
		uint AutoDeleteCheckIntervalSecs 300
		uint AutoSaveConfigSpan 300
		bool BackupConfigOnlyWhenModified true
		string CipherName AES128-SHA
		uint CurrentBuild 9731
		bool DisableCoreDumpOnUnix false
		bool DisableDeadLockCheck false
		bool DisableDosProction false
		bool DisableGetHostNameWhenAcceptTcp false
		bool DisableIntelAesAcceleration false
		bool DisableIPsecAggressiveMode false
		bool DisableIPv6Listener true
		bool DisableJsonRpcWebApi false
		bool DisableNatTraversal true
		bool DisableOpenVPNServer true
		bool DisableSessionReconnect false
		bool DisableSSTPServer false
		bool DontBackupConfig false
		bool EnableVpnAzure false
		bool EnableVpnOverDns false
		bool EnableVpnOverIcmp false
		byte HashedPassword {{.ServerHashedPassword}}
		string KeepConnectHost keepalive.softether.org
		uint KeepConnectInterval 50
		uint KeepConnectPort 80
		uint KeepConnectProtocol 1
		uint64 LoggerMaxLogSize 2
		uint MaxConcurrentDnsClientThreads 512
		uint MaxConnectionsPerIP 256
		uint MaxUnestablishedConnections 1000
		bool NoHighPriorityProcess false
		bool NoLinuxArpFilter false
		bool NoSendSignature false
		string OpenVPNDefaultClientOption dev-type$20tun,link-mtu$201500,tun-mtu$201500,cipher$20AES-128-CBC,auth$20SHA1,keysize$20128,key-method$202,tls-client
		string OpenVPN_UdpPortList 1194
		bool SaveDebugLog false
		uint ServerLogSwitchType 4
		uint ServerType 0
		bool StrictSyslogDatetimeFormat false
		bool Tls_Disable1_0 false
		bool Tls_Disable1_1 false
		bool Tls_Disable1_2 false
		bool UseKeepConnect true
		bool UseWebTimePage false
		bool UseWebUI false

		declare GlobalParams
		{
			uint FIFO_BUDGET 10240000
			uint HUB_ARP_SEND_INTERVAL 5000
			uint IP_TABLE_EXPIRE_TIME 60000
			uint IP_TABLE_EXPIRE_TIME_DHCP 300000
			uint MAC_TABLE_EXPIRE_TIME 600000
			uint MAX_BUFFERING_PACKET_SIZE 2560000
			uint MAX_HUB_LINKS 1024
			uint MAX_IP_TABLES 65536
			uint MAX_MAC_TABLES 65536
			uint MAX_SEND_SOCKET_QUEUE_NUM 128
			uint MAX_SEND_SOCKET_QUEUE_SIZE 2560000
			uint MAX_STORED_QUEUE_NUM 1024
			uint MEM_FIFO_REALLOC_MEM_SIZE 655360
			uint MIN_SEND_SOCKET_QUEUE_SIZE 320000
			uint QUEUE_BUDGET 2048
			uint SELECT_TIME 256
			uint SELECT_TIME_FOR_NAT 30
			uint STORM_CHECK_SPAN 500
			uint STORM_DISCARD_VALUE_END 1024
			uint STORM_DISCARD_VALUE_START 3
		}
		declare ServerTraffic
		{
			declare RecvTraffic
			{
				uint64 BroadcastBytes 0
				uint64 BroadcastCount 0
				uint64 UnicastBytes 0
				uint64 UnicastCount 0
			}
			declare SendTraffic
			{
				uint64 BroadcastBytes 0
				uint64 BroadcastCount 0
				uint64 UnicastBytes 0
				uint64 UnicastCount 0
			}
		}
		declare SyslogSettings
		{
			string HostName $
			uint Port 0
			uint SaveType 0
		}
	}
	declare VirtualHUB
	{
		declare DEFAULT
		{
			uint64 CreatedTime {{.CreatedTime}}
			byte HashedPassword {{.HubHashedPassword}}
			uint64 LastCommTime {{.CreatedTime}}
			uint64 LastLoginTime {{.CreatedTime}}
			uint NumLogin 0
			bool Online true
			bool RadiusConvertAllMsChapv2AuthRequestToEap false
			string RadiusRealm $
			uint RadiusRetryInterval 0
			uint RadiusServerPort 1812
			string RadiusSuffixFilter $
			bool RadiusUsePeapInsteadOfEap false
			byte SecurePassword {{.HubSecurePassword}}
			uint Type 0

			declare AccessList
			{
			}
			declare AdminOption
			{
				uint allow_hub_admin_change_option 0
				uint deny_bridge 0
				uint deny_change_user_password 0
				uint deny_empty_password 1
				uint deny_hub_admin_change_ext_option 0
				uint deny_qos 0
				uint deny_routing 0
				uint max_accesslists 0
				uint max_bitrates_download 0
				uint max_bitrates_upload 0
				uint max_groups 0
				uint max_multilogins_per_user 0
				uint max_sessions 0
				uint max_sessions_bridge 0
				uint max_sessions_client 0
				uint max_sessions_client_bridge_apply 0
				uint max_users 0
				uint no_access_list_include_file 0
				uint no_cascade 0
				uint no_change_access_control_list 0
				uint no_change_access_list 0
				uint no_change_admin_password 0
				uint no_change_cert_list 0
				uint no_change_crl_list 0
				uint no_change_groups 0
				uint no_change_log_config 0
				uint no_change_log_switch_type 0
				uint no_change_msg 0
				uint no_change_users 0
				uint no_delay_jitter_packet_loss 0
				uint no_delete_iptable 0
				uint no_delete_mactable 0
				uint no_disconnect_session 0
				uint no_enum_session 0
				uint no_offline 0
				uint no_online 0
				uint no_query_session 0
				uint no_read_log_file 0
				uint no_securenat 0
				uint no_securenat_enabledhcp 0
				uint no_securenat_enablenat 0
			}
			declare LogSetting
			{
				uint PacketLogSwitchType 4
				uint PACKET_LOG_ARP 0
				uint PACKET_LOG_DHCP 1
				uint PACKET_LOG_ETHERNET 0
				uint PACKET_LOG_ICMP 0
				uint PACKET_LOG_IP 0
				uint PACKET_LOG_TCP 0
				uint PACKET_LOG_TCP_CONN 1
				uint PACKET_LOG_UDP 0
				bool SavePacketLog false
				bool SaveSecurityLog false
				uint SecurityLogSwitchType 4
			}
			declare Message
			{
			}
			declare Option
			{
				uint AccessListIncludeFileCacheLifetime 30
				uint AdjustTcpMssValue 0
				bool ApplyIPv4AccessListOnArpPacket false
				bool AssignVLanIdByRadiusAttribute false
				bool BroadcastLimiterStrictMode false
				uint BroadcastStormDetectionThreshold 0
				uint ClientMinimumRequiredBuild 0
				bool DenyAllRadiusLoginWithNoVlanAssign false
				uint DetectDormantSessionInterval 0
				bool DisableAdjustTcpMss false
				bool DisableCheckMacOnLocalBridge false
				bool DisableCorrectIpOffloadChecksum false
				bool DisableHttpParsing false
				bool DisableIPParsing false
				bool DisableIpRawModeSecureNAT false
				bool DisableKernelModeSecureNAT false
				bool DisableUdpAcceleration false
				bool DisableUdpFilterForLocalBridgeNic false
				bool DisableUserModeSecureNAT false
				bool DoNotSaveHeavySecurityLogs false
				bool DropArpInPrivacyFilterMode true
				bool DropBroadcastsInPrivacyFilterMode true
				bool FilterBPDU false
				bool FilterIPv4 false
				bool FilterIPv6 false
				bool FilterNonIP false
				bool FilterOSPF false
				bool FilterPPPoE false
				uint FloodingSendQueueBufferQuota 33554432
				bool ManageOnlyLocalUnicastIPv6 true
				bool ManageOnlyPrivateIP true
				uint MaxLoggedPacketsPerMinute 0
				uint MaxSession 0
				bool NoArpPolling false
				bool NoDhcpPacketLogOutsideHub true
				bool NoEnum false
				bool NoIpTable false
				bool NoIPv4PacketLog false
				bool NoIPv6AddrPolling false
				bool NoIPv6DefaultRouterInRAWhenIPv6 true
				bool NoIPv6PacketLog false
				bool NoLookBPDUBridgeId false
				bool NoMacAddressLog true
				bool NoManageVlanId false
				bool NoPhysicalIPOnPacketLog false
				bool NoSpinLockForPacketDelay false
				bool RemoveDefGwOnDhcpForLocalhost true
				uint RequiredClientId 0
				uint SecureNAT_MaxDnsSessionsPerIp 0
				uint SecureNAT_MaxIcmpSessionsPerIp 0
				uint SecureNAT_MaxTcpSessionsPerIp 0
				uint SecureNAT_MaxTcpSynSentPerIp 0
				uint SecureNAT_MaxUdpSessionsPerIp 0
				bool SecureNAT_RandomizeAssignIp false
				bool SuppressClientUpdateNotification false
				bool UseHubNameAsDhcpUserClassOption false
				bool UseHubNameAsRadiusNasId false
				string VlanTypeId 0x8100
				bool YieldAfterStorePacket false
			}
			{{if .IsRelayNode}}
			declare CascadeList
			{
				declare Cascade0
				{
					bool CheckServerCert false
					bool Online true

					declare ClientAuth
					{
						uint AuthType 1
						byte HashedPassword {{.NextNodePassword}}
						string Username {{.NextNodeUserName}}
					}
					declare ClientOption
					{
						string AccountName {{.NextNodeName}}
						uint AdditionalConnectionInterval 1
						uint ConnectionDisconnectSpan 0
						string DeviceName _SEHUBLINKCLI_
						bool DisableQoS false
						bool HalfConnection false
						bool HideNicInfoWindow false
						bool HideStatusWindow false
						string Hostname {{.NextNodeHost}}
						string HubName DEFAULT
						uint MaxConnection 8
						bool NoRoutingTracking true
						bool NoTls1 false
						bool NoUdpAcceleration false
						uint NumRetry 4294967295
						uint Port {{.NextNodePort}}
						uint PortUDP 0
						string ProxyName $
						byte ProxyPassword $
						uint ProxyPort 0
						uint ProxyType 0
						string ProxyUsername $
						bool RequireBridgeRoutingMode true
						bool RequireMonitorMode false
						uint RetryInterval 10
						bool UseCompress false
						bool UseEncrypt true
					}
					declare Policy
					{
						bool ArpDhcpOnly false
						bool CheckIP false
						bool CheckIPv6 false
						bool CheckMac false
						bool DHCPFilter false
						bool DHCPForce false
						bool DHCPNoServer false
						bool DHCPv6Filter false
						bool DHCPv6NoServer false
						bool FilterIPv4 false
						bool FilterIPv6 false
						bool FilterNonIP false
						uint MaxDownload 0
						uint MaxIP 0
						uint MaxIPv6 0
						uint MaxMac 0
						uint MaxUpload 0
						bool NoBroadcastLimiter false
						bool NoIPv6DefaultRouterInRA false
						bool NoIPv6DefaultRouterInRAWhenIPv6 false
						bool NoServer false
						bool NoServerV6 false
						bool RAFilter false
						bool RSandRAFilter false
						uint VLanId 0
					}
				}
			}
			{{else}}
			declare SecureNAT
			{
				bool Disabled false
				bool SaveLog false

				declare VirtualDhcpServer
				{
					string DhcpDnsServerAddress 8.8.8.8
					string DhcpDnsServerAddress2 8.8.4.4
					string DhcpDomainName $
					bool DhcpEnabled true
					uint DhcpExpireTimeSpan 7200
					string DhcpGatewayAddress {{.Subnet}}.1
					string DhcpLeaseIPEnd {{.Subnet}}.20
					string DhcpLeaseIPStart {{.Subnet}}.10
					string DhcpPushRoutes $
					string DhcpSubnetMask 255.255.255.0
				}
				declare VirtualHost
				{
					string VirtualHostIp {{.Subnet}}.1
					string VirtualHostIpSubnetMask 255.255.255.0
					string VirtualHostMacAddress {{.VirtualHostMacAddress}}
				}
				declare VirtualRouter
				{
					bool NatEnabled true
					uint NatMtu 1500
					uint NatTcpTimeout 1800
					uint NatUdpTimeout 60
				}
			}
			{{end}}
			declare SecurityAccountDatabase
			{
				declare CertList
				{
				}
				declare CrlList
				{
				}
				declare GroupList
				{
				}
				declare IPAccessControlList
				{
				}
				declare UserList
				{
					{{range .UserList}}
					declare {{.UserName}}
					{
						byte AuthNtLmSecureHash {{.UserAuthNtLmSecureHash}}
						byte AuthPassword {{.UserAuthPassword}}
						uint AuthType 1
						uint64 CreatedTime {{$.CreatedTime}}
						uint64 ExpireTime 0
						uint64 LastLoginTime 0
						string Note $
						uint NumLogin 0
						string RealName $
						uint64 UpdatedTime {{$.CreatedTime}}

						declare Traffic
						{
							declare RecvTraffic
							{
								uint64 BroadcastBytes 0
								uint64 BroadcastCount 0
								uint64 UnicastBytes 0
								uint64 UnicastCount 0
							}
							declare SendTraffic
							{
								uint64 BroadcastBytes 0
								uint64 BroadcastCount 0
								uint64 UnicastBytes 0
								uint64 UnicastCount 0
							}
						}
					}
					{{end}}
				}
			}
			declare Traffic
			{
				declare RecvTraffic
				{
					uint64 BroadcastBytes 0
					uint64 BroadcastCount 0
					uint64 UnicastBytes 0
					uint64 UnicastCount 0
				}
				declare SendTraffic
				{
					uint64 BroadcastBytes 0
					uint64 BroadcastCount 0
					uint64 UnicastBytes 0
					uint64 UnicastCount 0
				}
			}
		}
	}
	declare VirtualLayer3SwitchList
	{
	}
}
`


src/sha/sha.go
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// Package sha implements the SHA-0 hash algorithm as defined in FIPSPUB 180.
//
// SHA-0 is cryptographically broken and should not be used for secure
// applications.
package sha

import (
	"crypto"
	"errors"
	"hash"
)

const SHA0 crypto.Hash = 100

func init() {
	//crypto.RegisterHash(SHA0, New)
}

// The size of a SHA-0 checksum in bytes.
const Size = 20

// The blocksize of SHA-0 in bytes.
const BlockSize = 64

const (
	chunk = 64
	init0 = 0x67452301
	init1 = 0xEFCDAB89
	init2 = 0x98BADCFE
	init3 = 0x10325476
	init4 = 0xC3D2E1F0
)

// digest represents the partial evaluation of a checksum.
type digest struct {
	h   [5]uint32
	x   [chunk]byte
	nx  int
	len uint64
}

const (
	magic         = "sha\x00"
	marshaledSize = len(magic) + 5*4 + chunk + 8
)

func (d *digest) MarshalBinary() ([]byte, error) {
	b := make([]byte, 0, marshaledSize)
	b = append(b, magic...)
	b = appendUint32(b, d.h[0])
	b = appendUint32(b, d.h[1])
	b = appendUint32(b, d.h[2])
	b = appendUint32(b, d.h[3])
	b = appendUint32(b, d.h[4])
	b = append(b, d.x[:d.nx]...)
	b = b[:len(b)+len(d.x)-int(d.nx)] // already zero
	b = appendUint64(b, d.len)
	return b, nil
}

func (d *digest) UnmarshalBinary(b []byte) error {
	if len(b) < len(magic) || string(b[:len(magic)]) != magic {
		return errors.New("my/sha: invalid hash state identifier")
	}
	if len(b) != marshaledSize {
		return errors.New("my/sha: invalid hash state size")
	}
	b = b[len(magic):]
	b, d.h[0] = consumeUint32(b)
	b, d.h[1] = consumeUint32(b)
	b, d.h[2] = consumeUint32(b)
	b, d.h[3] = consumeUint32(b)
	b, d.h[4] = consumeUint32(b)
	b = b[copy(d.x[:], b):]
	b, d.len = consumeUint64(b)
	d.nx = int(d.len) % chunk
	return nil
}

func appendUint64(b []byte, x uint64) []byte {
	var a [8]byte
	putUint64(a[:], x)
	return append(b, a[:]...)
}

func appendUint32(b []byte, x uint32) []byte {
	var a [4]byte
	putUint32(a[:], x)
	return append(b, a[:]...)
}

func consumeUint64(b []byte) ([]byte, uint64) {
	_ = b[7]
	x := uint64(b[7]) | uint64(b[6])<<8 | uint64(b[5])<<16 | uint64(b[4])<<24 |
		uint64(b[3])<<32 | uint64(b[2])<<40 | uint64(b[1])<<48 | uint64(b[0])<<56
	return b[8:], x
}

func consumeUint32(b []byte) ([]byte, uint32) {
	_ = b[3]
	x := uint32(b[3]) | uint32(b[2])<<8 | uint32(b[1])<<16 | uint32(b[0])<<24
	return b[4:], x
}

func (d *digest) Reset() {
	d.h[0] = init0
	d.h[1] = init1
	d.h[2] = init2
	d.h[3] = init3
	d.h[4] = init4
	d.nx = 0
	d.len = 0
}

// New returns a new hash.Hash computing the SHA1 checksum. The Hash also
// implements encoding.BinaryMarshaler and encoding.BinaryUnmarshaler to
// marshal and unmarshal the internal state of the hash.
func New() hash.Hash {
	d := new(digest)
	d.Reset()
	return d
}

func (d *digest) Size() int { return Size }

func (d *digest) BlockSize() int { return BlockSize }

func (d *digest) Write(p []byte) (nn int, err error) {
	nn = len(p)
	d.len += uint64(nn)
	if d.nx > 0 {
		n := copy(d.x[d.nx:], p)
		d.nx += n
		if d.nx == chunk {
			block(d, d.x[:])
			d.nx = 0
		}
		p = p[n:]
	}
	if len(p) >= chunk {
		n := len(p) &^ (chunk - 1)
		block(d, p[:n])
		p = p[n:]
	}
	if len(p) > 0 {
		d.nx = copy(d.x[:], p)
	}
	return
}

func (d *digest) Sum(in []byte) []byte {
	// Make a copy of d so that caller can keep writing and summing.
	d0 := *d
	hash := d0.checkSum()
	return append(in, hash[:]...)
}

func (d *digest) checkSum() [Size]byte {
	len := d.len
	// Padding.  Add a 1 bit and 0 bits until 56 bytes mod 64.
	var tmp [64]byte
	tmp[0] = 0x80
	if len%64 < 56 {
		d.Write(tmp[0 : 56-len%64])
	} else {
		d.Write(tmp[0 : 64+56-len%64])
	}

	// Length in bits.
	len <<= 3
	putUint64(tmp[:], len)
	d.Write(tmp[0:8])

	if d.nx != 0 {
		panic("d.nx != 0")
	}

	var digest [Size]byte

	putUint32(digest[0:], d.h[0])
	putUint32(digest[4:], d.h[1])
	putUint32(digest[8:], d.h[2])
	putUint32(digest[12:], d.h[3])
	putUint32(digest[16:], d.h[4])

	return digest
}

// ConstantTimeSum computes the same result of Sum() but in constant time
func (d *digest) ConstantTimeSum(in []byte) []byte {
	d0 := *d
	hash := d0.constSum()
	return append(in, hash[:]...)
}

func (d *digest) constSum() [Size]byte {
	var length [8]byte
	l := d.len << 3
	for i := uint(0); i < 8; i++ {
		length[i] = byte(l >> (56 - 8*i))
	}

	nx := byte(d.nx)
	t := nx - 56                 // if nx < 56 then the MSB of t is one
	mask1b := byte(int8(t) >> 7) // mask1b is 0xFF iff one block is enough

	separator := byte(0x80) // gets reset to 0x00 once used
	for i := byte(0); i < chunk; i++ {
		mask := byte(int8(i-nx) >> 7) // 0x00 after the end of data

		// if we reached the end of the data, replace with 0x80 or 0x00
		d.x[i] = (^mask & separator) | (mask & d.x[i])

		// zero the separator once used
		separator &= mask

		if i >= 56 {
			// we might have to write the length here if all fit in one block
			d.x[i] |= mask1b & length[i-56]
		}
	}

	// compress, and only keep the digest if all fit in one block
	block(d, d.x[:])

	var digest [Size]byte
	for i, s := range d.h {
		digest[i*4] = mask1b & byte(s>>24)
		digest[i*4+1] = mask1b & byte(s>>16)
		digest[i*4+2] = mask1b & byte(s>>8)
		digest[i*4+3] = mask1b & byte(s)
	}

	for i := byte(0); i < chunk; i++ {
		// second block, it's always past the end of data, might start with 0x80
		if i < 56 {
			d.x[i] = separator
			separator = 0
		} else {
			d.x[i] = length[i-56]
		}
	}

	// compress, and only keep the digest if we actually needed the second block
	block(d, d.x[:])

	for i, s := range d.h {
		digest[i*4] |= ^mask1b & byte(s>>24)
		digest[i*4+1] |= ^mask1b & byte(s>>16)
		digest[i*4+2] |= ^mask1b & byte(s>>8)
		digest[i*4+3] |= ^mask1b & byte(s)
	}

	return digest
}

// Sum returns the SHA-0 checksum of the data.
func Sum(data []byte) [Size]byte {
	var d digest
	d.Reset()
	d.Write(data)
	return d.checkSum()
}

func putUint64(x []byte, s uint64) {
	_ = x[7]
	x[0] = byte(s >> 56)
	x[1] = byte(s >> 48)
	x[2] = byte(s >> 40)
	x[3] = byte(s >> 32)
	x[4] = byte(s >> 24)
	x[5] = byte(s >> 16)
	x[6] = byte(s >> 8)
	x[7] = byte(s)
}

func putUint32(x []byte, s uint32) {
	_ = x[3]
	x[0] = byte(s >> 24)
	x[1] = byte(s >> 16)
	x[2] = byte(s >> 8)
	x[3] = byte(s)
}


src/sha/shablock.go
// Copyright 2009 The Go Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package sha

const (
	_K0 = 0x5A827999
	_K1 = 0x6ED9EBA1
	_K2 = 0x8F1BBCDC
	_K3 = 0xCA62C1D6
)

// blockGeneric is a portable, pure Go version of the SHA-0 block step.
func block(dig *digest, p []byte) {
	var w [16]uint32

	h0, h1, h2, h3, h4 := dig.h[0], dig.h[1], dig.h[2], dig.h[3], dig.h[4]
	for len(p) >= chunk {
		// Can interlace the computation of w with the
		// rounds below if needed for speed.
		for i := 0; i < 16; i++ {
			j := i * 4
			w[i] = uint32(p[j])<<24 | uint32(p[j+1])<<16 | uint32(p[j+2])<<8 | uint32(p[j+3])
		}

		a, b, c, d, e := h0, h1, h2, h3, h4

		// Each of the four 20-iteration rounds
		// differs only in the computation of f and
		// the choice of K (_K0, _K1, etc).
		i := 0
		for ; i < 16; i++ {
			f := b&c | (^b)&d
			a5 := a<<5 | a>>(32-5)
			b30 := b<<30 | b>>(32-30)
			t := a5 + f + e + w[i&0xf] + _K0
			a, b, c, d, e = t, a, b30, c, d
		}
		for ; i < 20; i++ {
			w[i&0xf] = w[(i-3)&0xf] ^ w[(i-8)&0xf] ^ w[(i-14)&0xf] ^ w[(i)&0xf]

			f := b&c | (^b)&d
			a5 := a<<5 | a>>(32-5)
			b30 := b<<30 | b>>(32-30)
			t := a5 + f + e + w[i&0xf] + _K0
			a, b, c, d, e = t, a, b30, c, d
		}
		for ; i < 40; i++ {
			w[i&0xf] = w[(i-3)&0xf] ^ w[(i-8)&0xf] ^ w[(i-14)&0xf] ^ w[(i)&0xf]
			f := b ^ c ^ d
			a5 := a<<5 | a>>(32-5)
			b30 := b<<30 | b>>(32-30)
			t := a5 + f + e + w[i&0xf] + _K1
			a, b, c, d, e = t, a, b30, c, d
		}
		for ; i < 60; i++ {
			w[i&0xf] = w[(i-3)&0xf] ^ w[(i-8)&0xf] ^ w[(i-14)&0xf] ^ w[(i)&0xf]
			f := ((b | c) & d) | (b & c)

			a5 := a<<5 | a>>(32-5)
			b30 := b<<30 | b>>(32-30)
			t := a5 + f + e + w[i&0xf] + _K2
			a, b, c, d, e = t, a, b30, c, d
		}
		for ; i < 80; i++ {
			w[i&0xf] = w[(i-3)&0xf] ^ w[(i-8)&0xf] ^ w[(i-14)&0xf] ^ w[(i)&0xf]
			f := b ^ c ^ d
			a5 := a<<5 | a>>(32-5)
			b30 := b<<30 | b>>(32-30)
			t := a5 + f + e + w[i&0xf] + _K3
			a, b, c, d, e = t, a, b30, c, d
		}

		h0 += a
		h1 += b
		h2 += c
		h3 += d
		h4 += e

		p = p[chunk:]
	}

	dig.h[0], dig.h[1], dig.h[2], dig.h[3], dig.h[4] = h0, h1, h2, h3, h4
}


src/main.go
package main

import (
	"encoding/base64"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"text/template"
	"time"

	"./sha"
	"golang.org/x/crypto/md4"
)

type user struct {
	UserName               string
	UserAuthNtLmSecureHash string //generateNtPasswordHash(pass)
	UserAuthPassword       string //hashPassword(UserName, pass)
}

type params struct {
	DDnsClientKey        string //hashAdminPassword(random)
	ServerHashedPassword string //hashAdminPassword(random)
	HubHashedPassword    string //hashAdminPassword(hub_random)
	CreatedTime          int64
	HubSecurePassword    string //hashPassword("administrator", hub_random)

	UserList []user

	IsRelayNode      bool
	NextNodeName     string
	NextNodeHost     string
	NextNodePort     uint64
	NextNodeUserName string
	NextNodePassword string

	Subnet                string
	VirtualHostMacAddress string
}

func hashPassword(name, pass string) string {
	sha := sha.New()
	sha.Write([]byte(pass))
	sha.Write([]byte(strings.ToUpper(name)))
	return base64.StdEncoding.EncodeToString(sha.Sum(nil))
}

func generateNtPasswordHash(pass string) string {
	md4 := md4.New()
	var buf = make([]byte, len(pass)*2)
	for i, c := range pass {
		buf[i*2] = byte(c)
	}
	md4.Write(buf)
	return base64.StdEncoding.EncodeToString(md4.Sum(nil))
}

func hashAdminPassword(pass string) string {
	sha := sha.New()
	sha.Write([]byte(pass))
	return base64.StdEncoding.EncodeToString(sha.Sum(nil))
}

func generateConfigFile(p *params) {
	t := template.Must(template.New("config").Parse(configTemplate))
	out, err := os.Create("./vpn_server.config")
	if err != nil {
		fmt.Printf("[ERR] Create config file: %v\n", err)
		return
	}
	defer out.Close()
	t.Execute(out, p)
}

func randString(n int) string {
	var buf []byte = make([]byte, n)
	rand.Read(buf)
	for i, b := range buf {
		buf[i] = ' ' + (b % ('~' - ' '))
	}
	return string(buf)
}
func randLowerLetter(n int) string {
	var buf []byte = make([]byte, n)
	rand.Read(buf)
	for i, b := range buf {
		buf[i] = 'a' + (b % ('z' - 'a'))
	}
	return string(buf)
}
func splitBySemicolon(c rune) bool {
	return c == ';'
}

func main() {
	var p params
	var mac []byte = []byte{5: 0}
	var pass string
	p.CreatedTime = time.Now().Unix()
	rand.Seed(p.CreatedTime)
	pass = randString(8)
	p.DDnsClientKey = hashAdminPassword(pass)

	pass = randString(16)
	p.ServerHashedPassword = hashAdminPassword(pass)

	pass = randString(12)
	p.HubHashedPassword = hashAdminPassword(pass)
	p.HubSecurePassword = hashPassword("administrator", pass)

	rand.Read(mac)
	mac[0] &= 0xFE
	p.VirtualHostMacAddress = fmt.Sprintf("%02X-%02X-%02X-%02X-%02X-%02X", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5])
	p.Subnet = fmt.Sprintf("192.168.%d", (rand.Uint32() % 256))

	var userList = strings.FieldsFunc(os.Getenv("UserList"), splitBySemicolon)

	_, p.IsRelayNode = os.LookupEnv("RelayNode")
	if p.IsRelayNode {
		p.NextNodeHost = os.Getenv("NextNodeHost")
		p.NextNodePort, _ = strconv.ParseUint(os.Getenv("NextNodePort"), 10, 16)
		p.NextNodeUserName = os.Getenv("NextNodeUserName")
		p.NextNodePassword = hashPassword(p.NextNodeUserName, os.Getenv("NextNodePassword"))
		p.NextNodeName = randLowerLetter(6)
	}

	if len(userList) == 0 {
		p.UserList = make([]user, 1)
		p.UserList[0].UserName = randLowerLetter(6)
		pass = randString(8)
		p.UserList[0].UserAuthPassword = hashPassword(p.UserList[0].UserName, pass)
		p.UserList[0].UserAuthNtLmSecureHash = generateNtPasswordHash(pass)
		fmt.Printf("[INFO] Random generated user is : %s\nPassword is %s\n", p.UserList[0].UserName, pass)
	} else {
		p.UserList = make([]user, len(userList))
		for i, u := range userList {
			var temp = strings.SplitN(u, ":", 2)
			if len(temp) == 1 {
				p.UserList[i].UserName = temp[0]
				pass = randString(8)
				p.UserList[i].UserAuthPassword = hashPassword(temp[0], pass)
				p.UserList[i].UserAuthNtLmSecureHash = generateNtPasswordHash(pass)
				fmt.Printf("[INFO] Random generated password for user '%s' is: %s\n", temp[0], pass)
			} else {
				p.UserList[i].UserName = temp[0]
				p.UserList[i].UserAuthPassword = hashPassword(temp[0], temp[1])
				p.UserList[i].UserAuthNtLmSecureHash = generateNtPasswordHash(temp[1])
			}
		}
	}

	generateConfigFile(&p)
}
